### PR TITLE
[SPARK-28301][SQL] fix the behavior of table name resolution with multi-catalog

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Now users can register multiple catalogs in Spark, and the table name resolution should be compatible with multi-catalog. The expected behavior is simple:
* For DDL commands that can only deal with tables
    * If the table name has only one name part, then it's a table in the default catalog.
    * If the table name has more than one name part like `a.b.c`.
        * if `a` is a registered catalog, then it's a table `c` under namespace `b` in catalog `a`.
        * if `a` is not a registered catalog, then it's a table `c` under namespace `a.b` in the default catalog.
* For SELECT/INSERT that can handle both tables and temp views, first check if the table name is a temp view or global temp view, otherwise the rule is the same as DDL commands.

However, we need to change the expected behavior a little bit because the builtin hive catalog hasn't migrated to the new catalog API yet:
1. If the default catalog config is set, pick it as the default catalog. Otherwise pick hive catalog as the default catalog.
2. If the default catalog config is not set, and the table name has more than 2 name parts. We should fail with "no catalog specified for table"

The current behavior of table name resolution is a little confusing:
* For DDL commands that can only deal with tables
    * If the first part of the table name matches a registered catalog, then it's a table in that catalog. (expected)
    * Otherwise, if the table name has less than 3 parts, and the provider name is v1, go with the builtin Hive catalog. (This is not expected. By design different catalogs can interprete table provider name differently. We should go with the default catalog if the config is set, no matter what the table provider name is.)
* For SELECT/INSERT that can handle both tables and temp views
    * If the first part of the table name matches a registered catalog, then it's a table in that catalog. (expected)
    * Otherwise, if the table name has less than 3 parts, go with the builtin Hive catalog. (This is not expected as we need to respect the default catalog config.)
    * Otherwise, the query is unresolved. (This is not expected as we need to respect the default catalog config.)

This PR fixes the behavior of the table name resolution:
1. Now `Analyzer` instead of `SparkSession` tracks all the registered catalogs. This is because `SparkSession` is in sql/core not sql/catalyst.
2. The rule `DataSourceResolution` only resolves table name to the hive catalog when the catalog is not specified in the table name and default catalog is not set.

## How was this patch tested?

new test cases

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -752,29 +732,21 @@ class Analyzer(
...
       try {
-        catalog.lookupRelation(tableIdentWithDb)
+        catalog.lookupRelation(identWithDb)
       } catch {
         case _: NoSuchTableException | _: NoSuchDatabaseException =>
           u
       }
     }
-
-    // If the database part is specified, and we support running SQL directly on files, and
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: This is not needed anymore because in https://github.com/apache/spark/pull/24741 we delay the error reporting of unresolved relation to `CheckAnalysis`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala [hunk 2]
```diff
@@ -21,70 +21,96 @@ import scala.collection.JavaConverters._
...
     spark.sql("DROP TABLE source")
     spark.sql("DROP TABLE source2")
   }
 
-  test("CreateTable: use v2 plan because catalog is set") {
-    spark.sql("CREATE TABLE testcat.table_name (id bigint, data string) USING foo")
-
-    val testCatalog = spark.catalog("testcat").asTableCatalog
-    val table = testCatalog.loadTable(Identifier.of(Array(), "table_name"))
+  test("CreateTable: basic") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/v2/DataSourceV2SQLSuite.scala]: This test well demonstrates the expected behavior after this fix.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
cc @rxin  @marmbrus @rdblue @dongjoon-hyun  @gengliangwang

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @cloud-fan . The failure seems to be the same one.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by rdblue** [comment]
-1

I agree with the idea behind this PR, which is that the v2 catalog default should be applied the same way in all cases... that's why I fixed this in #24768. From that PR's description:

> Updates catalog lookup to always provide the default if it is set for consistent behavior.

And from discussion:

> After this PR, the rules to determine the catalog responsible for an identifier are:
> 1. If the identifier starts with a known catalog, use it
> 2. If there is a configured default v2 catalog, use that catalog
> 3. Otherwise, the session catalog is responsible for the identifier

While this PR includes some additional fixes for handling for temporary tables, I don't understand why this was necessary when another PR is proposing the same behavior change and could be updated to fix those issues as well.

This alternative implementation reverses choices that were intended to minimize the risk to existing users of adding multi-catalog support to Spark 3.0. That plan is to change as few parts of v1 as possible to ensure its behavior does not change, but this PR makes the v1 `SessionCatalog` responsible for proxying calls to load v2 tables by modifying the signature and behavior of `lookupRelation`. These changes are not needed to update the default catalog behavior.

Similarly, `AsTableIdentifier` is used in rules that previously matched `TableIdentifier` and guarantees that the rule will not accidentally match an identifier with a catalog part. But this PR rewrites `AsTableIdentifier` so that it will match any 2-part identifier without the catalog check so it now must be applied after `CatalogObjectIdentifier`. Previously correct uses of `AsTableIdentifier` are now unsafe.

Last, this PR makes changes that are unrelated to fixing the behavior of the default catalog. For example, this adds a `CatalogManager`, which may be a good idea, but should be proposed and added separately.

To sum it up, this appears to be proposing a different implementation, not different behavior. But it is quite invasive and different from the design that we have been building to up to now. **Some of these changes may be good ideas, but let's separate them into their own PRs and discuss the merits of each**. If you want to change the design to add a proxy catalog, then please bring it up at the sync or write up a proposal.

**Comment by cloud-fan** [comment]
The reasons that I open this PR:
1. The fix in #24768 leads to different behaviors than I expect: a) temp view should have higher priority. b) table provider should not affect table name resolution. The implementation also needs to be different for the proposed behaviors, that's one reason I open this PR.
2. #24768 does 2 things together: 1. fix the table name resolution. 2. support v2 provider with Hive catalog. Since discussion is needed for 1, I think it's more efficient to have an individual PR for it. Since the implementation in my mind is very different, I open a new PR instead of asking you to separate the changes to a new PR.

> but this PR makes the v1 SessionCatalog responsible for proxying calls to load v2 tables by modifying the signature and behavior of lookupRelation.

The `lookupRelation` should belong to analyzer instead of catalog, since catalog doesn't know the relation plans Spark uses. We put it in `SessionCatalog` for historical reasons, I can move it to analyzer in a new PR if it's better. **To clarify, I agree with you that v1 session catalog should not proxy calls to v2 catalogs.**

>  But this PR rewrites AsTableIdentifier so that it will match any 2-part identifier without the catalog check so it now must be applied after CatalogObjectIdentifier.

Thanks for catching this bug! I feel it's cleaner to call `AsTableIdentifier` after `CatalogObjectIdentifier`, but we must guarantee that when `AsTableIdentifier` is called, no v2 catalog can be found. I've pushed a commit to fix it.

Note that, the main point of this PR is to trigger discussion and reach a consensus about the behavior of table name resolution, including all the details. Once we reach a consensus, we can merge your PR first as long as it matches the expected behavior (you can take some tests in this PR), and rebase my PR as a refactor.

**Comment by cloud-fan** [comment]
To avoid future confusion, I've reverted the changes in `SessionCatalog` and moved the related changes to `Analyzer`.

**Comment by cloud-fan** [comment]
retest this please

**Comment by rdblue** [comment]
> The fix in #24768 leads to different behaviors than I expect: a) temp view should have higher priority. b) table provider should not affect table name resolution. The implementation also needs to be different for the proposed behaviors, that's one reason I open this PR.

Usually, I would expect you to point out problems with a PR in a review instead of opening a separate PR that doesn't mention the original. In the future, just coordinate about how to get the work done.

> 1. fix the table name resolution. 2. support v2 provider with Hive catalog

That PR solves these two at the same time because the default catalog was introduced so that v2 providers could be used. If the default catalog behavior changes, then the v2 session catalog needs to be introduced to handle those cases. Similarly, if we add the v2 session catalog, then we need to update the table resolution rules.

It is reasonable to fix those two at the same time. Fixing just table resolution in this PR breaks v2 providers, which is unnecessary.

> The lookupRelation should belong to analyzer instead of catalog, since catalog doesn't know the relation plans Spark uses

I'm not sure what you mean. The `ResolveTable` rule already does this in the analyzer. Moving this to `SessionCatalog` is a different approach entirely. Like I said, I'm happy to discuss the merits of this, but please don't mix these changes together.

> the main point of this PR is to trigger discussion and reach a consensus about the behavior of table name resolution

This doesn't make sense to me because we agree about what the behavior should be. This doesn't propose substantial changes to the behavior in the other PR, just minor fixes to temporary table handling that we all agree on. In practice, this PR only introduces an alternative implementation that breaks existing conventions and approaches.

**Comment by cloud-fan** [comment]
> It is reasonable to fix those two at the same time. Fixing just table resolution in this PR breaks v2 providers

Have we ever supported v2 provider in Hive catalog before?

I think one major argument here is: can we separate "fix the table name resolution" and "support v2 provider with Hive catalog"? I don't see a strong reason to do them together and I open this PR to show that it's possible to only fix table name resolution.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 3]
```diff
@@ -181,29 +181,32 @@ case class CreateViewCommand(
...
+              s"referencing a temporary view $tblIdent")
+          }
         case other if !other.resolved => other.expressions.flatMap(_.collect {
           // Disallow creating permanent views based on temporary UDFs.
           case e: UnresolvedFunction
             if sparkSession.sessionState.catalog.isTemporaryFunction(e.name) =>
             throw new AnalysisException(s"Not allowed to create a permanent view $name by " +
               s"referencing a temporary function `${e.name}`")
         })
+        case _ =>
```
Comment (albertusk95) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: I think it'd be better to add a comment like `do nothing`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 4]
```diff
@@ -181,29 +181,32 @@ case class CreateViewCommand(
    * Permanent views are not allowed to reference temp objects, including temp function and views
    */
   private def verifyTemporaryObjectsNotExists(sparkSession: SparkSession): Unit = {
```
Comment (albertusk95) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: for the sake of consistency, I think it'd be better to use either `sparkSession` or `session` like the below methods after this one

**Review by albertusk95 — CHANGES_REQUESTED** [review]
